File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        total = 1
        zero_count = 0

        for i in range(len(nums)):
            if (nums[i] == 0):
                zero_count += 1
                continue
            total *= nums[i]

        result = []
        if (zero_count >= 2):
            for i in range(len(nums)):
                result.append(0)
            return result

        for i in range(len(nums)):
            if (nums[i] == 0):
                result.append(total)
            elif (zero_count == 1):
                result.append(0)
            else :
                result.append(int(total/nums[i]))

        return result
        

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        
        Map<Integer, Integer> cumProduct = new HashMap<Integer, Integer>();
        
        int product = 1;
        for (int i = 0; i<n; i++){
            product *= nums[i];
            cumProduct.put(i, product);
        }

        int[] result = new int[n];
        result[n-1] = cumProduct.get(n-2);

        product = nums[n-1];        
        for (int i = n-2; i>0; i--){
            result[i] = cumProduct.get(i-1) * product;
            product *= nums[i];
        }
        result[0] = product;

        return result;
    }
}